lucene/core/src/java/org/apache/lucene/geo/EdgeTree.java

```diff
@@ -90,30 +103,34 @@
   // THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
   // CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
   // IN THE SOFTWARE.
-  protected boolean contains(double x, double y, AtomicBoolean isOnEdge) {
-    boolean res = false;
-    if (isOnEdge.get() == false && y <= this.max) {
+  private byte containsPnPoly(double x, double y) {
+    byte res = FALSE;
+    if (y <= this.max) {
       if (y == this.y1 && y == this.y2 ||
           (y <= this.y1 && y >= this.y2) != (y >= this.y1 && y <= this.y2)) {
         if ((x == this.x1 && x == this.x2) ||
             ((x <= this.x1 && x >= this.x2) != (x >= this.x1 && x <= this.x2) &&
                 GeoUtils.orient(this.x1, this.y1, this.x2, this.y2, x, y) == 0)) {
-          // if its on the boundary return true
-          isOnEdge.set(true);
-          return true;
+          return ON_EDGE;
         } else if (this.y1 > y != this.y2 > y) {
-          res = x < (this.x2 - this.x1) * (y - this.y1) / (this.y2 - this.y1) + this.x1;
+          res = x < (this.x2 - this.x1) * (y - this.y1) / (this.y2 - this.y1) + this.x1 ? TRUE : FALSE;
         }
       }
       if (this.left != null) {
-        res ^= left.contains(x, y, isOnEdge);
+        res ^= left.containsPnPoly(x, y);
+        if ((res & 0x02) == 0x02) {
+          return ON_EDGE;
+        }
       }
 
       if (this.right != null && y >= this.low) {
-        res ^= right.contains(x, y, isOnEdge);
+        res ^= right.containsPnPoly(x, y);
+        if ((res & 0x02) == 0x02) {
+          return ON_EDGE;
+        }
       }
     }
-    return isOnEdge.get() || res;
+    return res;
```

Comment:
let's assert that the return value is in the expected range for sanity?
```suggestion
    assert res >= FALSE && res <= EN_EDGE;
    return res;
```